cpp/src/groupby/new_groupby.cu

```diff
@@ -219,3 +226,140 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
+
+  return std::make_tuple(unique_indices, new_sz);
+}
+
+std::tuple<cudf::table, gdf_index_type*, gdf_size_type> 
+gdf_group_by_without_aggregations(cudf::table const& input_table,
+                                  gdf_size_type num_key_cols,
+                                  gdf_index_type const * key_col_indices,
+                                  gdf_context* context)
+{
+  CUDF_EXPECTS(nullptr != key_col_indices, "key_col_indices is null");
+  CUDF_EXPECTS(0 < num_key_cols, "number of key colums should be greater than zero");
+
+  if (0 == input_table.num_rows()) {
+    return std::make_tuple(cudf::table(), nullptr, 0);
+  }
+
+  gdf_size_type nrows = input_table.num_rows();
+  
+  // Allocate output columns
+  std::vector<gdf_dtype> output_types(input_table.num_columns());
+  for(gdf_size_type i = 0; i < input_table.num_columns(); ++i) {
+    auto input_col = input_table.get_column(i);
+    output_types[i] = input_col->dtype;
+  }  
+  cudf::table destination_table(nrows, output_types, true);
+
+  bool group_by_keys_contain_nulls = false;
+  std::vector<gdf_column*> key_cols_vect(num_key_cols);
+  for (gdf_size_type i = 0; i < num_key_cols; i++){
+    key_cols_vect[i] = const_cast<gdf_column*>(input_table.get_column(i));
+    group_by_keys_contain_nulls = (group_by_keys_contain_nulls || key_cols_vect[i]->null_count > 0);
+  }
+  cudf::table key_col_table(key_cols_vect.data(), key_cols_vect.size());
+
+  rmm::device_vector<gdf_size_type> sorted_indices(nrows);
+  gdf_column sorted_indices_col;
+  CUDF_TRY(gdf_column_view(&sorted_indices_col, (void*)(sorted_indices.data().get()),
+                          nullptr, nrows, GDF_INT32));
+
+  if (context->flag_groupby_include_nulls || !group_by_keys_contain_nulls){  // SQL style
+    CUDF_TRY(gdf_order_by(key_col_table.begin(),
+                          nullptr,
+                          key_col_table.num_columns(),
+                          &sorted_indices_col,
+                          context));
+  } else {  // Pandas style
+    gdf_context temp_ctx;
+    temp_ctx.flag_null_sort_behavior = GDF_NULL_AS_LARGEST_FOR_MULTISORT;
+
+    CUDF_TRY(gdf_order_by(key_col_table.begin(),
+                          nullptr,
+                          key_col_table.num_columns(),
+                          &sorted_indices_col,
+                          &temp_ctx));
+
+    // lets filter out all the nulls in the group by key column by:
+    // we will take the data which has been sorted such that the nulls in the group by keys are all last
+    // then using row_bitmask we can count how many rows have a null in the group by keys and use that 
+    // to resize the data
+    auto orderby_cols_bitmask = row_bitmask(key_col_table);
+    int valid_count;
+    CUDF_TRY(gdf_count_nonzero_mask(reinterpret_cast<gdf_valid_type*>(orderby_cols_bitmask.data().get()),
+                                    nrows, &valid_count));
+    
+    for (gdf_size_type i = 0; i < input_table.num_columns(); ++i) {
```

Comment:
Picking a nit here, but it's generally nicer to use an algorithm rather than a "raw" for loop. For example, this could be replaced by:
```
std::transform(destination_table.begin(), destination_table.begin(), 
                       [valid_count](gdf_column * col){col->size = valid_count;});
```

Reply:
done